web3indexer/processor.py

```diff
@@ -0,0 +1,468 @@
+
+import structlog
+from web3 import Web3
+
+from .constants import (
+    ERC721_TRANSFER_TOPIC,
+    ERC1155_TRANSFER_SINGLE_TOPIC,
+    ERC1155_TRANSFER_BATCH_TOPIC,
+    ERC_721_IDENTIFIER,
+    ERC_721_METADATA_IDENTIFIER,
+    ERC_1155_IDENTIFIER,
+    ERC_1155_METADATA_IDENTIFIER,
+)
+from .crud import (
+    get_contract,
+    get_nft,
+    upsert_contract,
+    upsert_nft,
+    upsert_transfer,
+)
+from .models import Contract, ContractType, Nft, Transfer
+from .task import ProcessBlockTask
+from .utils import get_nft_id, read_file
+
+
+logger = structlog.get_logger()
+
+ERC165_ABI = read_file("abi/ERC165.json")
+ERC1155_ABI = read_file("abi/ERC1155.json")
+ERC721_ABI = read_file("abi/ERC721.json")
+
+
+class BlockProcessor:
+    """
+    Handles fetching, parsing and storing information for a given block.
+    """
+
+    MAX_RETRIES = 5
+
+    def __init__(self, db):
+        self.db = db
+
+    def process_with_retry(self, dispatcher, w3, task):
+        try:
+            self.process(dispatcher, w3, task)
+        except Exception as e:
+            logger.error(e)
+            if task.retries < self.MAX_RETRIES:
+                dispatcher.put(ProcessBlockTask(block_number=task.block_number))
+            else:
+                raise Exception(
+                    "Reached max number of retries for processing block number {}".format(
+                        task.block_number
+                    )
+                )
+
+    def process(self, dispatcher, w3, task):
+        logger.info("Processing block", block_number=task.block_number)
+
+        block = w3.eth.get_block(task.block_number)
+        timestamp = datetime.fromtimestamp(block.timestamp)
+
+        for transaction in block.transactions:
+            txn_receipt = w3.eth.get_transaction_receipt(transaction)
+
+            for log_index, log in enumerate(txn_receipt.logs):
+                self._process_log(w3, log, log_index, timestamp)
+
+    def _process_log(self, w3: Web3, log, log_index: int, timestamp: datetime):
+        topics = log.topics
+        contract_address = log.address
+        event_signature = topics[0].hex()
+        if (
+            event_signature == ERC721_TRANSFER_TOPIC
+            and self._supports_interface(
+                w3, contract_address, ERC_721_IDENTIFIER
+            )
+        ):
+            self._process_erc721_log(w3, log, log_index, timestamp)
+        elif (
+            event_signature == ERC1155_TRANSFER_SINGLE_TOPIC
+            or event_signature == ERC1155_TRANSFER_BATCH_TOPIC
+        ) and self._supports_interface(
+            w3, contract_address, ERC_1155_IDENTIFIER
+        ):
+            if event_signature == ERC1155_TRANSFER_SINGLE_TOPIC:
+                self._process_erc1155_transfer_single_log(
+                    w3, log, log_index, timestamp
+                )
+            elif event_signature == ERC1155_TRANSFER_BATCH_TOPIC:
+                self._process_erc1155_transfer_batch_log(
```

Comment:
On first thought this seems like a good place to enqueue a new task to do this processing rather than do it in the same task that fetches the block. That was the processing of each log can be parallelized rather than needing to be done in sequence.